File: 139. Word Break.cpp
class Solution {
public:
    // // brute force: match starting from head, and go on
    // // TLE for large data
    // bool wordBreak(string s, vector<string>& wordDict) {
    //     if(wordDict.empty() || s.empty()) return false;
    //     for(int i = 0; i < wordDict.size(); i++){
    //         if(s.find(wordDict[i]) == 0){
    //             if(wordDict[i].size() == s.size())
    //                 return true;
    //             string sub = s.substr(wordDict[i].size());
    //             if(wordBreak(sub, wordDict))
    //                 return true;
    //         }
    //     }
    //     return false;
    // }

    // dp: dp[i] means substring with len i starting from head is a valid one
    // O(mn)time, O(m)space, m:length of s, n:length of wordDict, 88.68%, ok
    bool wordBreak(string s, vector<string>& wordDict) {
        if(wordDict.empty() || s.empty()) return false;
        int len = s.size();
        vector<int> dp(len + 1, 0);
        dp[0] = 1;
        for(int i = 1; i <= len; i++){
            for(auto w: wordDict){
                int head = i - w.size();
                if(head >= 0 && dp[head] && s.substr(head, i - head) == w){
                    dp[i] = 1;
                    break;
                }
            }
        }
        return dp[len];
    }
};

File: 139. Word Break.py
class Solution(object):
    
    # backtracking: O(N^N) time, O(N) space, TLE
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        if s == '':
            return True
        for w in wordDict:
            if len(w) <= len(s) and w == s[:len(w)]:
                if self.wordBreak(s[len(w):], wordDict):
                    return True
        return False
    
    # !! backtracking with memorization: O(N2) time, O(N) space, TLE
    def wordBreak(self, s, wordDict):
        def helper(s, start, wordDict):
            if start == len(s):
                return True
            if self.memo.get(start, None) is not None:
                return self.memo[start]
            for end in xrange(start, len(s)):
                if s[start:end + 1] in wordDict and helper(s, end + 1, wordDict):
                    self.memo[start] = True
                    return True
            self.memo[start] = False
            return False
        self.memo = {}
        return helper(s, 0, wordDict)
    
    # dynamic programming: O(N2) time, O(N) space
    def wordBreak(self, s, wordDict):
        # dp[l]: whether substr from head (length l) is word compound
        dp = [False for i in range(len(s) + 1)]   
        dp[0] = True   # base: empty string is True
        for l in xrange(1, len(s) + 1):
            for i in xrange(l):
                if dp[i] and s[i: l] in wordDict:
                    dp[l] = True
        return dp[-1]
    
    # BFS
        